src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,225 +285,409 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
+    /// request was handled (typically by returning an error message).  If <see langword="false"/>
+    /// is returned, the request is processed normally.
+    /// <br/><br/>
+    /// By default this does not check authorization rules becuase authentication may take place within
+    /// the WebSocket connection during the ConnectionInit message.  Authorization checks for
+    /// WebSocket connections occur then, after authorization has taken place.
+    /// </summary>
+    protected virtual ValueTask<bool> HandleAuthorizeWebSocketConnectionAsync(HttpContext context, RequestDelegate next)
+        => new(false);
+
+    /// <summary>
+    /// Handles a single GraphQL request.
+    /// </summary>
     protected virtual async Task HandleRequestAsync(
         HttpContext context,
         RequestDelegate next,
-        IDictionary<string, object> userContext,
-        IList<GraphQLRequest> bodyGQLBatchRequest,
-        GraphQLRequest gqlRequest,
-        IDocumentExecuter<TSchema> executer,
-        CancellationToken cancellationToken)
+        GraphQLRequest gqlRequest)
     {
         // Normal execution with single graphql request
-        if (bodyGQLBatchRequest == null)
-        {
-            var stopwatch = ValueStopwatch.StartNew();
-            await RequestExecutingAsync(gqlRequest);
-            var result = await ExecuteRequestAsync(gqlRequest, userContext, executer, context.RequestServices, cancellationToken);
-
-            await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequest, result, stopwatch.Elapsed));
+        var userContext = await BuildUserContextAsync(context);
+        var result = await ExecuteRequestAsync(context, gqlRequest, context.RequestServices, userContext);
+        var statusCode = Options.ValidationErrorsReturnBadRequest && !result.Executed
+            ? HttpStatusCode.BadRequest
+            : HttpStatusCode.OK;
+        await WriteJsonResponseAsync(context, statusCode, result);
+    }
 
-            await WriteResponseAsync(context.Response, _serializer, cancellationToken, result);
+    /// <summary>
+    /// Handles a batched GraphQL request.
+    /// </summary>
+    protected virtual async Task HandleBatchRequestAsync(
+        HttpContext context,
+        RequestDelegate next,
+        IList<GraphQLRequest> gqlRequests)
+    {
+        var userContext = await BuildUserContextAsync(context);
+        var results = new ExecutionResult[gqlRequests.Count];
+        if (gqlRequests.Count == 1)
+        {
+            results[0] = await ExecuteRequestAsync(context, gqlRequests[0], context.RequestServices, userContext);
         }
-        // Execute multiple graphql requests in one batch
         else
         {
-            var executionResults = new ExecutionResult[bodyGQLBatchRequest.Count];
-            for (int i = 0; i < bodyGQLBatchRequest.Count; ++i)
+            // Batched execution with multiple graphql requests
+            if (!Options.BatchedRequestsExecuteInParallel)
```

Comment:
This option works for all types of queries enabling executing them in parallel. I am afraid that in the case of mutations, the result will be unforesewable and this is not what is expected.

Reply:
Probably the option should be disabled by default.  I don't use batched requests but what you say makes sense.

Reply:
Yes, definetely disabled by default + warn about mutations in comments. Ideally I would check operation type of all requests in batch and run in parallel this batch if and only if all requests in batch are queries. But this implies that requests should be parsed 😞 .Also I think that someone may want to run requests in parallel even for mutations so this behavior may be configurable. Perhaps the boolean option is enough for now.

In general, we should find what the spec says (http one or apollo) about executing requests in batches and follow it.

Reply:
> follow it

Perhaps “follow it by default”, allowing configuration. I would think that setting this option might be nearly as common as changing a query to use serial execution - for people that use batched requests.

Reply:
I want to read some docs regarding batching tomorrow.

Reply:
Excellent. Please let me know what you find and post some links if applicable.

Reply:
Today I spent time reading the GraphQL over HTTP spec (again). I found some batching documentation to read tomorrow, but it seems to me that when I did it last time, then nothing was said there about the parallelism of the queries.

Reply:
One new note - docs mention that Apollo Client handles batching using apollo-link-batch-http. In our case I doubt that we can do the same with new middleware design since `IConfigureExecution` returns single `ExecutionResult`.

Reply:
🤦 it's about client, not server

Reply:
> Request batching is prone to slow loading times on the client. For example, if five requests from separate components are made to the GraphQL endpoint, and one of the five requests take a long time, the client won’t get any of the results back until all operations have been resolved. In other words, batched operations are always **as slow as the slowest operation in the batch.**


So I can conclude that requests in bacht at least can be run in parallel.

Reply:
Parallel execution is enabled by default in the new options class, but can be configured.

https://github.com/graphql-dotnet/server/pull/774/files#diff-b8a172e1e8e08ba82d01aa83350e5fa2e32f1a40beee713891b05366db1fed5aR45